File: ball_stencil/meshcheck.py
"""Mesh validation: watertight manifold, winding, radius, triangle quality."""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np


@dataclass
class MeshReport:
    n_vertices: int
    n_faces: int
    is_watertight: bool
    is_manifold: bool
    consistent_winding: bool
    n_boundary_edges: int
    n_nonmanifold_edges: int
    max_radius_error_mm: float
    edge_len_min: float
    edge_len_max: float
    edge_len_mean: float
    max_aspect_ratio: float
    signed_volume_mm3: float
    n_degenerate: int

    def ok(self, cfg) -> bool:
        return (
            self.is_watertight
            and self.is_manifold
            and self.consistent_winding
            and self.max_radius_error_mm <= cfg.radius_tolerance_mm
            and self.n_degenerate == 0
        )

    def format(self, cfg) -> str:
        status = "PASS" if self.ok(cfg) else "FAIL"
        lines = [
            f"  manifold check ............ {status}",
            f"  vertices / faces .......... {self.n_vertices} / {self.n_faces}",
            f"  watertight ................ {self.is_watertight}",
            f"  manifold (edge deg==2) .... {self.is_manifold} "
            f"(boundary={self.n_boundary_edges}, nonmanifold={self.n_nonmanifold_edges})",
            f"  consistent winding ........ {self.consistent_winding}",
            f"  max |‖P‖-R| ............... {self.max_radius_error_mm:.2e} mm "
            f"(tol {cfg.radius_tolerance_mm})",
            f"  edge length min/mean/max .. {self.edge_len_min:.2f} / "
            f"{self.edge_len_mean:.2f} / {self.edge_len_max:.2f} mm",
            f"  max aspect ratio .......... {self.max_aspect_ratio:.2f}",
            f"  degenerate triangles ...... {self.n_degenerate}",
            f"  signed volume ............. {self.signed_volume_mm3:.1f} mm^3",
        ]
        return "\n".join(lines)


def check_mesh(mesh, cfg) -> MeshReport:
    v = mesh.vertices
    f = mesh.faces
    n = mesh.n_planar

    # --- edges ---------------------------------------------------------------
    directed = np.vstack([f[:, [0, 1]], f[:, [1, 2]], f[:, [2, 0]]])
    undirected = np.sort(directed, axis=1)
    uniq, counts = np.unique(undirected, axis=0, return_counts=True)
    n_boundary = int(np.sum(counts == 1))
    n_nonmanifold = int(np.sum(counts > 2))
    # An empty mesh must not pass: np.all([]) is True and sums over empty
    # arrays are 0, which would otherwise report watertight/manifold/clean.
    has_faces = len(f) > 0
    is_watertight = has_faces and n_boundary == 0 and n_nonmanifold == 0
    is_manifold = bool(has_faces and np.all(counts == 2))

    # consistent winding: each directed edge appears at most once
    du = np.unique(directed, axis=0, return_counts=True)[1]
    consistent = bool(np.all(du == 1)) and is_watertight

    # --- radius error --------------------------------------------------------
    norms = np.linalg.norm(v, axis=1)
    err_outer = np.abs(norms[:n] - cfg.outer_radius_mm)
    err_inner = np.abs(norms[n:] - cfg.inner_radius_mm)
    max_radius_error = float(max(err_outer.max(initial=0.0), err_inner.max(initial=0.0)))

    # --- triangle quality ----------------------------------------------------
    a, b, c = v[f[:, 0]], v[f[:, 1]], v[f[:, 2]]
    e0 = np.linalg.norm(b - a, axis=1)
    e1 = np.linalg.norm(c - b, axis=1)
    e2 = np.linalg.norm(a - c, axis=1)
    edges = np.concatenate([e0, e1, e2])
    longest = np.maximum.reduce([e0, e1, e2])
    shortest = np.minimum.reduce([e0, e1, e2])
    with np.errstate(divide="ignore", invalid="ignore"):
        aspect = np.where(shortest > 1e-12, longest / shortest, np.inf)
    # Degeneracy by AREA, not just shortest edge: a near-collinear sliver can
    # have three non-trivial edges yet ~zero area and an ill-defined normal.
    areas = 0.5 * np.linalg.norm(np.cross(b - a, c - a), axis=1)
    n_degenerate = int(np.sum((areas <= 1e-9) | (shortest <= 1e-12)))
    max_aspect = float(np.max(aspect[np.isfinite(aspect)], initial=0.0))

    # --- signed volume -------------------------------------------------------
    vol = float(np.sum(a * np.cross(b, c)) / 6.0)

    return MeshReport(
        n_vertices=len(v),
        n_faces=len(f),
        is_watertight=is_watertight,
        is_manifold=is_manifold,
        consistent_winding=consistent,
        n_boundary_edges=n_boundary,
        n_nonmanifold_edges=n_nonmanifold,
        max_radius_error_mm=max_radius_error,
        edge_len_min=float(edges.min(initial=0.0)),
        edge_len_max=float(edges.max(initial=0.0)),
        edge_len_mean=float(edges.mean()) if len(edges) else 0.0,
        max_aspect_ratio=max_aspect,
        signed_volume_mm3=vol,
        n_degenerate=n_degenerate,
    )

File: web/src/pipeline/meshcheck.ts
/**
 * Mesh validation: watertight manifold, winding, radius, triangle quality.
 * Port of ball_stencil/meshcheck.py.
 */

import { Mesh } from "./meshbuild";
import { Params, innerRadius, outerRadius } from "./config";

export interface MeshReport {
  nVertices: number;
  nFaces: number;
  isWatertight: boolean;
  isManifold: boolean;
  consistentWinding: boolean;
  nBoundaryEdges: number;
  nNonmanifoldEdges: number;
  maxRadiusErrorMm: number;
  edgeLenMin: number;
  edgeLenMax: number;
  edgeLenMean: number;
  maxAspectRatio: number;
  signedVolumeMm3: number;
  nDegenerate: number;
}

export function reportOk(r: MeshReport, p: Params): boolean {
  return (
    r.isWatertight &&
    r.isManifold &&
    r.consistentWinding &&
    r.maxRadiusErrorMm <= p.radius_tolerance_mm &&
    r.nDegenerate === 0
  );
}

export function checkMesh(mesh: Mesh, p: Params): MeshReport {
  const v = mesh.vertices;
  const f = mesh.faces;
  const n = mesh.nPlanar;
  const nFaces = f.length / 3;
  const nVertices = v.length / 3;
  const base = nVertices; // edge-key base

  // --- edges ---
  const undCount = new Map<number, number>();
  const dirSeen = new Set<number>();
  let dirDup = false;
  const addUnd = (a: number, b: number) => {
    const k = a < b ? a * base + b : b * base + a;
    undCount.set(k, (undCount.get(k) ?? 0) + 1);
  };
  const addDir = (a: number, b: number) => {
    const k = a * base + b;
    if (dirSeen.has(k)) dirDup = true;
    else dirSeen.add(k);
  };
  for (let i = 0; i < nFaces; i++) {
    const a = f[i * 3], b = f[i * 3 + 1], c = f[i * 3 + 2];
    addUnd(a, b); addUnd(b, c); addUnd(c, a);
    addDir(a, b); addDir(b, c); addDir(c, a);
  }
  let nBoundary = 0;
  let nNonmanifold = 0;
  let allDeg2 = true;
  for (const cnt of undCount.values()) {
    if (cnt === 1) nBoundary++;
    if (cnt > 2) nNonmanifold++;
    if (cnt !== 2) allDeg2 = false;
  }
  const hasFaces = nFaces > 0;
  const isWatertight = hasFaces && nBoundary === 0 && nNonmanifold === 0;
  const isManifold = hasFaces && allDeg2;
  const consistentWinding = hasFaces && !dirDup && isWatertight;

  // --- radius error ---
  let maxErr = 0;
  const outerR = outerRadius(p);
  const innerR = innerRadius(p);
  for (let i = 0; i < n; i++) {
    const norm = Math.hypot(v[i * 3], v[i * 3 + 1], v[i * 3 + 2]);
    const e = Math.abs(norm - outerR);
    if (e > maxErr) maxErr = e;
  }
  for (let i = n; i < nVertices; i++) {
    const norm = Math.hypot(v[i * 3], v[i * 3 + 1], v[i * 3 + 2]);
    const e = Math.abs(norm - innerR);
    if (e > maxErr) maxErr = e;
  }

  // --- triangle quality + volume ---
  let edgeMin = Infinity, edgeMax = 0, edgeSum = 0;
  let maxAspect = 0;
  let nDegenerate = 0;
  let vol6 = 0;
  for (let i = 0; i < nFaces; i++) {
    const ia = f[i * 3], ib = f[i * 3 + 1], ic = f[i * 3 + 2];
    const ax = v[ia * 3], ay = v[ia * 3 + 1], az = v[ia * 3 + 2];
    const bx = v[ib * 3], by = v[ib * 3 + 1], bz = v[ib * 3 + 2];
    const cx = v[ic * 3], cy = v[ic * 3 + 1], cz = v[ic * 3 + 2];
    const e0 = Math.hypot(bx - ax, by - ay, bz - az);
    const e1 = Math.hypot(cx - bx, cy - by, cz - bz);
    const e2 = Math.hypot(ax - cx, ay - cy, az - cz);
    edgeSum += e0 + e1 + e2;
    edgeMin = Math.min(edgeMin, e0, e1, e2);
    edgeMax = Math.max(edgeMax, e0, e1, e2);
    const longest = Math.max(e0, e1, e2);
    const shortest = Math.min(e0, e1, e2);
    if (shortest > 1e-12) {
      const asp = longest / shortest;
      if (asp > maxAspect) maxAspect = asp;
    }
    // area = 0.5 |(b-a) x (c-a)|
    const ux = bx - ax, uy = by - ay, uz = bz - az;
    const wx = cx - ax, wy = cy - ay, wz = cz - az;
    const crx = uy * wz - uz * wy;
    const cry = uz * wx - ux * wz;
    const crz = ux * wy - uy * wx;
    const area = 0.5 * Math.hypot(crx, cry, crz);
    if (area <= 1e-9 || shortest <= 1e-12) nDegenerate++;
    // a . (b x c)
    const bcx = by * cz - bz * cy;
    const bcy = bz * cx - bx * cz;
    const bcz = bx * cy - by * cx;
    vol6 += ax * bcx + ay * bcy + az * bcz;
  }
  const nEdges = nFaces * 3;

  return {
    nVertices,
    nFaces,
    isWatertight,
    isManifold,
    consistentWinding,
    nBoundaryEdges: nBoundary,
    nNonmanifoldEdges: nNonmanifold,
    maxRadiusErrorMm: maxErr,
    edgeLenMin: nEdges ? edgeMin : 0,
    edgeLenMax: nEdges ? edgeMax : 0,
    edgeLenMean: nEdges ? edgeSum / nEdges : 0,
    maxAspectRatio: maxAspect,
    signedVolumeMm3: vol6 / 6.0,
    nDegenerate,
  };
}
